gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1150 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+\********************************************************************/
+
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+#include <vector>
+#include <string>
+#include <numeric>
+#include <algorithm>
+
+extern "C" {
+#include "Transaction.h"
+#include "engine-helpers.h"
+#include "dialog-utils.h"
+#include "assistant-stock-transaction.h"
+#include "gnc-account-sel.h"
+#include "gnc-amount-edit.h"
+#include "gnc-component-manager.h"
+#include "gnc-date-edit.h"
+#include "gnc-tree-view-account.h"
+
+static QofLogModule log_module = GNC_MOD_ASSISTANT;
+
+void stock_assistant_prepare (GtkAssistant  *assistant, GtkWidget *page,
+                              gpointer user_data);
+void stock_assistant_finish  (GtkAssistant *assistant, gpointer user_data);
+void stock_assistant_cancel  (GtkAssistant *gtkassistant, gpointer user_data);
+}
+
+static const char* ASSISTANT_STOCK_TRANSACTION_CM_CLASS = "assistant-stock-transaction";
+
+enum assistant_pages
+{
+    PAGE_INTRO = 0,
+    PAGE_TRANSACTION_TYPE,
+    PAGE_TRANSATION_DETAILS,
+    PAGE_STOCK_AMOUNT,
+    PAGE_STOCK_VALUE,
+    PAGE_CASH,
+    PAGE_FEES,
+    PAGE_DIVIDEND,
+    PAGE_CAPGAINS,
+    PAGE_FINISH
+};
+
+enum split_cols
+{
+    SPLIT_COL_ACCOUNT = 0,
+    SPLIT_COL_MEMO,
+    SPLIT_COL_DEBIT,
+    SPLIT_COL_CREDIT,
+    NUM_SPLIT_COLS
+};
+
+/** structures *********************************************************/
+
+enum FieldMask
+{
+    DISABLED = 0,
+    ENABLED_DEBIT,
+    ENABLED_CREDIT,
+    ALLOW_ZERO = 4,
+    ALLOW_NEGATIVE = 8
+};
+
+struct TxnTypeInfo
+{
+    uint stock_amount;
+    uint stock_value;
+    uint cash_value;
+    uint fees_value;
+    bool fees_capitalize;
+    uint dividend_value;
+    uint capgains_value;
+    const char* friendly_name;
+    const char* explanation;
+};
+
+using StringVec = std::vector<std::string>;
+using TxnTypeVec = std::vector<TxnTypeInfo>;
+
+static const TxnTypeVec starting_types
+{
+    {
+        ENABLED_DEBIT,          // stock_amt
+        ENABLED_DEBIT,          // stock_val
+        ENABLED_CREDIT,         // cash_amt
+        ENABLED_DEBIT | ALLOW_ZERO,          // fees_amt
+        true,                   // fees_capitalize
+        DISABLED,               // dividend_amt
+        DISABLED,               // capg_amt
+        _("Open buy"),
+        _("Initial stock purchase")
```

Comment:
Static variables are created at program initialization before `main` so there's no locale or gettext text_domain. That means that these strings won't get translated. The fix is to use the `N_(x)` macro to mark them for xgettext and then call `gettext` (via the `_(x)` macro) when you pass them to the GtkWidget for display.

Short strings tend to be a bit cryptic in the isolation of the po file and the Weblate interface so it's helpful to translators to provide context. For that you use `NC_(x)` and `C_(x)`, and the context string goes first, e.g.:
```
NC_("Stock assistant open a position with a buy", "Open buy"),
```
and
```
          gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (info->transaction_type_combo),
                                             C_(it.friendly_name));
```

Reply:
> Short strings tend to be a bit cryptic in the isolation of the po file and the Weblate interface so it's helpful to translators to provide context. For that you use `NC_(x)` and `C_(x)`, and the context string goes first, e.g.:

Doesn't quite work like this. Both `NC_` and `C_` require both context & string -- https://people.gnome.org/~ryanl/glib-docs/glib-I18N.html -- Perhaps it's best use `Translators:` comment to provide explanation.

Reply:
Translator comments are a poor substitute because two instance of the same short English phrase sometimes has different translations in some language. While xgettext will copy `Translators` comments into gnucash.pot next to the `msgid`s without using context it will combine the same phrase into a single `msgid` so that the translator can provide only a single translation.

Reply:
I'm still confused about NC_ and C_ -- both require context and translatable string. I guess for this file the context will be "StockAssistant" throughout this file? i.e. `NC("StockAssistant", "Cover Buy")`, and `C_("StockAssistant", it.friendly_name)`?

Reply:
From https://people.gnome.org/~ryanl/glib-docs/glib-I18N.html#C-:CAPS using `C_` requires 2 args to be string _literals_. Maybe https://docs.gtk.org/glib/func.dpgettext2.html is the right one to use?

Reply:
We're mostly using gettext directly so you should consult it's documentaion; for this purpose that's [Using Contexts to Solve Ambiguities](https://www.gnu.org/software/gettext/manual/gettext.html#Contexts) and [Special Cases](https://www.gnu.org/software/gettext/manual/gettext.html#Special-cases). Note at the bottom of the Contexts section there's `pgettext_expr` that takes non-literal arguments.

Yes, you must repeat the context. `NC_` marks the string pair for xgettext to extract and returns only the second string, the msgid, so you have to repeat the context in calling `C_`.

Of course we can simplify that with our own set of macros (untested, you might need to play with it a bit):
```
struct context_msgid
{
    const char* context;
    const char* msgid;
};

#undef NC_
#undef CN_

#define NC_(c, m) \
  context_msgid(c, m)

#define CN_(m) \
   pgettext_expr (m.context, m.msgid)
```

We might want that in libgnucash/core-utils/gnc-locale-utils.h. The only minor issue is that the header declaring `pgettext_expr` is installed at $PREFIX/share/gettext/gettext.h, a rather strange place that will require adding a `-I` compiler argument to the relevant CMakeLists.txt.